cmd/release/main.go

```diff
@@ -0,0 +1,583 @@
+package main
+
+import (
+	"archive/zip"
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"flag"
+	"fmt"
+	"io"
+	"io/fs"
+	"log"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"aead.dev/minisign"
+	"github.com/bufbuild/buf/private/bufpkg/bufplugin/bufpluginref"
+	"github.com/bufbuild/plugins/internal/plugin"
+	"github.com/google/go-github/v48/github"
+	"github.com/hashicorp/go-retryablehttp"
+	"golang.org/x/mod/semver"
+	"golang.org/x/oauth2"
+)
+
+const (
+	githubOwner = "bufbuild"
+	// This is separate from githubOwner for testing releases (can point at personal fork)
+	githubReleaseOwner   = githubOwner // "pkwarren"
+	githubRepo           = "plugins"
+	packageTypeContainer = "container"
+	pluginReleasesFile   = "plugin-releases.json"
+)
+
+type ReleaseStatus int
+
+const (
+	EXISTING ReleaseStatus = iota
+	NEW
+	UPDATED
+)
+
+type PluginReleases struct {
+	Releases []PluginRelease `json:"releases"`
+}
+
+type PluginRelease struct {
+	PluginName       string        `json:"name"`              // org/name for the plugin (without remote)
+	PluginVersion    string        `json:"version"`           // version of the plugin (including 'v' prefix)
+	PluginZipDigest  string        `json:"zip_digest"`        // <digest-type>:<digest> for plugin .zip download
+	PluginYAMLDigest string        `json:"yaml_digest"`       // <digest-type>:<digest> for buf.plugin.yaml
+	GHCRImageDigest  string        `json:"ghcr_image_digest"` // <digest-type>:<digest> for ghcr.io/bufbuild/plugins-<org>-<name>:v<version> image
+	ReleaseTag       string        `json:"release_tag"`       // GitHub release tag - i.e. 20221121.1
+	URL              string        `json:"url"`               // URL to GitHub release zip file for the plugin - i.e. https://github.com/bufbuild/plugins/releases/download/20221121.1/bufbuild-connect-go-v1.1.0.zip
+	LastUpdated      time.Time     `json:"last_updated"`
+	Status           ReleaseStatus `json:"-"`
+}
+
+type pluginNameVersion struct {
+	name, version string
+}
+
+func main() {
+	var (
+		minisignPrivateKey string
+		dryRun             bool
+	)
+	flag.BoolVar(&dryRun, "dry-run", false, "perform a dry-run (no GitHub modifications)")
+	flag.StringVar(&minisignPrivateKey, "minisign-private-key", "", "path to minisign private key file")
+	flag.Parse()
+
+	if len(flag.Args()) != 1 {
+		_, _ = fmt.Fprintf(os.Stderr, "usage: %s [-dry-run] [-minisign-private-key <file>] <directory>\n", os.Args)
+		os.Exit(2)
+	}
+	root := flag.Args()[0]
+	if err := run(root, minisignPrivateKey, dryRun); err != nil {
+		log.Fatalln(err.Error())
```

Comment:
This is the only fatal logger - this ensures we clean things up (like temporary directories) on error.